### PR TITLE
chore: return something when ctx is done

### PR DESCRIPTION
**Pull request by eliasscosta**
# Why

Try to catch the memory leak.

Screens from grmon. 

image

With the new code:

image

### CODE REVIEW

**Review by math3vz — APPROVED** [review]

File: pipeline/backend/kubernetes/kubernetes.go [hunk 1]
```diff
@@ -231,43 +233,34 @@ func (e *kube) WaitStep(ctx context.Context, step *types.Step, taskUUID string)
...
+		if pod.Name != podName {
+			return
+		}
 
-			switch pod.Status.Phase {
-			case v1.PodSucceeded, v1.PodFailed, v1.PodUnknown:
-				finished <- true
-			}
+		if isImagePullBackOffState(pod) || isCompleted(pod) || pod.Status.Phase == v1.PodSucceeded || pod.Status.Phase == v1.PodFailed || pod.Status.Phase == v1.PodUnknown {
+			finished <- true
```
Comment (fernandrone) [inline pipeline/backend/kubernetes/kubernetes.go:241]: Are these changes from line 236 - 241 necessary? I feel they don't really introduce any change, they only reorganize the code.

Comment (fernandrone) [reply pipeline/backend/kubernetes/kubernetes.go:241]: I think this is fine for testing, but I would remove this changes so we don't have to come back and undo them later.

File: pipeline/backend/kubernetes/kubernetes.go [hunk 2]
```diff
@@ -231,43 +233,34 @@ func (e *kube) WaitStep(ctx context.Context, step *types.Step, taskUUID string)
...
 		return nil, err
 	}
 
 	stop := make(chan struct{})
-	si.Start(stop)
 	defer close(stop)
 
-	// TODO Cancel on ctx.Done
-	<-finished
+	go podInformer.Run(stop)
```
Comment (fernandrone) [inline pipeline/backend/kubernetes/kubernetes.go:257]: What's the difference between podInformer.Run and si.Start ? Is this what fixes the problem?

File: pipeline/backend/kubernetes/kubernetes.go [hunk 3]
```diff
@@ -231,43 +233,34 @@ func (e *kube) WaitStep(ctx context.Context, step *types.Step, taskUUID string)
...
-	last := <-finished
-	log.Trace().
-		Str("taskUUID", taskUUID).
-		Str("step", step.Name).
-		Msgf("Pod has: %v", last)
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case <-finished:
+	}
```
Comment (fernandrone) [inline pipeline/backend/kubernetes/kubernetes.go:263]: That's nice 👍🏻

File: pipeline/backend/kubernetes/kubernetes.go [hunk 4]
```diff
@@ -288,13 +281,11 @@ func (e *kube) WaitStep(ctx context.Context, step *types.Step, taskUUID string)
...
 
-	bs := &types.State{
+	return &types.State{
 		ExitCode:  int(cs.State.Terminated.ExitCode),
 		Exited:    true,
 		OOMKilled: false,
-	}
-
-	return bs, nil
+	}, nil
```
Comment (fernandrone) [inline pipeline/backend/kubernetes/kubernetes.go:288]: I think this is a slight improvement to the code, however this is not necessary either. If we try to add this change upstream someone from the Woodpecker team might complain that it's unnecessary. I would remove it.

**Pull request closed without merging**